py/objtype.c

```diff
@@ -1015,6 +1015,9 @@ STATIC void type_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
             // Returns a read-only dict of the class attributes.
             // If the internal locals is not fixed, a copy will be created.
             mp_obj_dict_t *dict = self->locals_dict;
+            if (!dict) {
+                dict = MP_OBJ_TO_PTR(mp_obj_new_dict(0));
+            }
```

Comment:
Would it be better/faster to make a dict from `mp_const_empty_map` to ensure it's read-only / fixed?

Reply:
> Would it be better/faster to make a dict from mp_const_empty_map to ensure it's read-only / fixed?

No, because is the fails at the next step. e.g. dir(zip) works, but zip.__dict__ fails. The Suggested change of @jimmo is fine in both cases.

Reply:
The problem with this fix is that it'll create two dicts: one here and then a copy of this empty dict just below.  So it's inefficient from that point of view.

Might be an idea to create a `mp_const_empty_dict`, and to keep code size down then redefine `mp_const_empty_map` like this:
```c
#define mp_const_empty_map (&mp_const_empty_dict_obj.map)
```

Reply:
OK good idea.

Done, but without the address-of to match the existing uses... i.e.

```c
#define mp_const_empty_map (mp_const_empty_dict_obj.map)
```